File: backend/src/services/adminAlunoService.js
const { IllegalArgumentError } = require('../middleware/errorHandler');
const { rolesEquivalentes } = require('../utils/roleUtils');

class AdminAlunoService {
  constructor(firestoreService, firebaseUserProvisioningService) {
    this.firestoreService = firestoreService;
    this.firebaseUserProvisioningService = firebaseUserProvisioningService;
  }

  async listar() {
    const docs = await this.firestoreService.listarUsuariosPorRole('ALUNO');
    return Promise.all(docs.map((d) => this.paraAlunoAdmin(d)));
  }

  async buscarPorId(id) {
    const doc = await this.firestoreService.buscarUsuarioPorId(id);
    if (!doc) return null;
    return this.paraAlunoAdmin(doc);
  }

  async criar(nome, email, matricula, curso) {
    this.validarDados(nome, email, matricula, curso);
    const uid = await this.firebaseUserProvisioningService.upsertUser(null, email, nome, matricula);
    await this.firestoreService.salvarUsuario(uid, await this.payloadAluno(uid, nome, email, matricula, curso, null));
    return await this.buscarPorId(uid);
  }

  async atualizar(id, nome, email, matricula, curso) {
    this.validarDados(nome, email, matricula, curso);
    const alunoExistente = await this.firestoreService.buscarUsuarioPorId(id);
    if (!alunoExistente) return null;
    if (!rolesEquivalentes(alunoExistente.get('role'), 'ALUNO')) return null;

    await this.firebaseUserProvisioningService.upsertUser(id, email, nome, matricula);
    await this.firestoreService.salvarUsuario(id, await this.payloadAluno(id, nome, email, matricula, curso, alunoExistente));
    return await this.buscarPorId(id);
  }

  async remover(id) {
    const atual = await this.buscarPorId(id);
    if (!atual) return false;
    await this.firebaseUserProvisioningService.deleteByUid(id);
    await this.firestoreService.removerUsuario(id);
    return true;
  }

  validarDados(nome, email, matricula, curso) {
    if (this.isBlank(nome) || this.isBlank(email) || this.isBlank(matricula) || this.isBlank(this.cursoId(curso))) {
      throw new IllegalArgumentError('Todos os campos sao obrigatorios.');
    }
    if (String(matricula).trim().length < 6) {
      throw new IllegalArgumentError('A matricula precisa ter ao menos 6 caracteres.');
    }
  }

  async payloadAluno(uid, nome, email, matricula, curso, atual) {
    const payload = {
      uid,
      nome: String(nome).trim(),
      email: String(email).trim().toLowerCase(),
      role: 'ALUNO'
    };

    const vinculoAtual = atual ? this.mapVinculo(atual.get('vinculo')) : null;

    const vinculo = {
      id_curso: this.cursoId(curso),
      id_turma: vinculoAtual ? this.texto(vinculoAtual.id_turma) : '',
      matricula: String(matricula).trim(),
      ch_total_exigida: 200,
      status_no_curso: 'Ativo',
      saldos: { ensino: 0, pesquisa: 0, extensao: 0 }
    };

    payload.vinculo = vinculo;
    return payload;
  }

  mapVinculo(rawVinculo) {
    if (!rawVinculo) return null;
    if (typeof rawVinculo === 'object' && !Array.isArray(rawVinculo)) return { ...rawVinculo };
    if (Array.isArray(rawVinculo) && rawVinculo.length > 0 && typeof rawVinculo[0] === 'object') return { ...rawVinculo[0] };
    return null;
  }

  async paraAlunoAdmin(doc) {
    const aluno = {
      id: doc.id,
      nome: this.texto(doc.get('nome')),
      email: this.texto(doc.get('email'))
    };

    const vinculo = this.mapVinculo(doc.get('vinculo'));
    if (vinculo) {
      aluno.matricula = this.texto(vinculo.matricula);
      aluno.curso = await this.cursoDoVinculo(vinculo);
      aluno.turma = this.texto(vinculo.id_turma);
    }
    return aluno;
  }

  async cursoDoVinculo(vinculo) {
    const cursoRaw = vinculo.id_curso || vinculo.curso;
    const nomeNoVinculo = this.cursoNome(cursoRaw);
    if (!this.isBlank(nomeNoVinculo)) return nomeNoVinculo;

    const idCurso = this.cursoId(cursoRaw);
    const cursoDoc = await this.firestoreService.buscarCurso(idCurso);
    const nomeCurso = cursoDoc ? this.texto(cursoDoc.get('nome_curso')) : '';
    return this.isBlank(nomeCurso) ? idCurso : nomeCurso;
  }

  cursoId(value) {
    if (value == null) return '';
    if (typeof value === 'object') {
      if (Array.isArray(value)) return value.length ? this.cursoId(value[0]) : '';
      return this.texto(value.id_curso || value.idCurso || value.id || value.codigo || value.codigoCurso).trim();
    }
    return this.texto(value).trim();
  }

  cursoNome(value) {
    if (value == null || typeof value !== 'object') return '';
    if (Array.isArray(value)) return value.length ? this.cursoNome(value[0]) : '';
    return this.texto(value.nome_curso || value.nomeCurso || value.nome || value.descricao);
  }

  isBlank(v) {
    return v == null || String(v).trim().length === 0;
  }

  texto(v) {
    return v == null ? '' : String(v);
  }
}

module.exports = AdminAlunoService;

File: backend/app/services/admin_aluno_service.rb
# frozen_string_literal: true

class AdminAlunoService
  def initialize(
    firestore_service: FirestoreService.new,
    firebase_auth_client: FirebaseAuthClient.new
  )
    @firestore = firestore_service
    @firebase_auth = firebase_auth_client
  end

  def listar
    @firestore.listar_usuarios_por_role("ALUNO").map { |doc| para_aluno_admin(doc) }
  end

  def buscar_por_id(id)
    doc = @firestore.buscar_usuario_por_id(id)
    return nil unless doc
    return nil unless FirestoreKeyNormalizer.role_matches?(role_do_usuario(doc), "aluno")

    para_aluno_admin(doc)
  end

  def criar(nome:, email:, matricula:, curso:)
    validar_dados(nome, email, matricula, curso)
    uid = @firebase_auth.upsert_user(
      preferred_uid: nil,
      email: email,
      display_name: nome,
      password: matricula
    )
    @firestore.salvar_usuario(uid, payload_aluno(uid, nome, email, matricula, curso, nil))
    buscar_por_id(uid) || raise(ArgumentError, "Falha ao criar aluno.")
  end

  def atualizar(id, nome:, email:, matricula:, curso:)
    validar_dados(nome, email, matricula, curso)
    existente = @firestore.buscar_usuario_por_id(id)
    return nil unless existente
    return nil unless FirestoreKeyNormalizer.role_matches?(role_do_usuario(existente), "aluno")

    @firebase_auth.upsert_user(
      preferred_uid: id,
      email: email,
      display_name: nome,
      password: matricula
    )
    @firestore.salvar_usuario(id, payload_aluno(id, nome, email, matricula, curso, existente))
    buscar_por_id(id)
  end

  def remover(id)
    return false unless buscar_por_id(id)

    @firebase_auth.delete_by_uid(id)
    @firestore.remover_usuario(id)
    true
  end

  private

  def validar_dados(nome, email, matricula, curso)
    raise ArgumentError, "Todos os campos sao obrigatorios." if [nome, email, matricula, curso].any?(&:blank?)
    raise ArgumentError, "A matricula precisa ter ao menos 6 caracteres." if matricula.to_s.strip.length < 6
  end

  def para_aluno_admin(doc)
    data = doc[:data]
    vinculo = map_vinculo(FirestoreKeyNormalizer.fetch_field(data, "vinculo"))

    if vinculo
      {
        id: doc[:id],
        nome: FirestoreKeyNormalizer.fetch_field(data, "nome"),
        email: FirestoreKeyNormalizer.fetch_field(data, "email"),
        matricula: texto(vinculo["matricula"]),
        curso: texto(vinculo["id_curso"]),
        turma: texto(vinculo["id_turma"])
      }
    else
      {
        id: doc[:id],
        nome: FirestoreKeyNormalizer.fetch_field(data, "nome"),
        email: FirestoreKeyNormalizer.fetch_field(data, "email"),
        matricula: FirestoreKeyNormalizer.fetch_field(data, "matricula"),
        curso: texto(FirestoreKeyNormalizer.fetch_field(data, "id_curso", "curso")),
        turma: texto(FirestoreKeyNormalizer.fetch_field(data, "id_turma", "turma"))
      }
    end
  end

  def payload_aluno(uid, nome, email, matricula, curso, atual)
    vinculo_atual = atual ? map_vinculo(FirestoreKeyNormalizer.fetch_field(atual[:data], "vinculo")) : nil
    id_curso_norm = FirestoreKeyNormalizer.normalize(curso)

    vinculo = {
      "id_curso" => id_curso_norm.presence || curso.to_s.strip,
      "id_turma" => vinculo_atual ? texto(vinculo_atual["id_turma"]) : "",
      "matricula" => matricula.to_s.strip,
      "ch_total_exigida" => 200,
      "status_no_curso" => "ativo",
      "saldos" => {
        "ensino" => 0,
        "pesquisa" => 0,
        "extensao" => 0
      }
    }

    {
      "uid" => uid,
      "nome" => nome.to_s.strip,
      "email" => email.to_s.strip.downcase,
      "role" => "aluno",
      "vinculo" => vinculo
    }
  end

  def map_vinculo(raw_vinculo)
    return stringify_hash(raw_vinculo) if raw_vinculo.is_a?(Hash)
    return stringify_hash(raw_vinculo.first) if raw_vinculo.is_a?(Array) && raw_vinculo.first.is_a?(Hash)

    nil
  end

  def stringify_hash(hash)
    hash.each_with_object({}) { |(k, v), h| h[k.to_s] = v }
  end

  def role_do_usuario(doc)
    FirestoreKeyNormalizer.fetch_field(doc[:data], "role", "perfil", "tipo")
  end

  def texto(value)
    value.nil? ? "" : value.to_s
  end
end
